### ISSUE TITLE
Nested List Support

### ISSUE THREAD

**Issue by toshimaru** [ijjk/notion-blog#4]
List on notion-blog is not nested correctly.

Any plan to support a nested list ❓ 

## List on Notion

![image](https://user-images.githubusercontent.com/803398/72956493-3d16db00-3de3-11ea-8e8a-7df353e6c40c.png)

## List on notion-blog

![image](https://user-images.githubusercontent.com/803398/72956401-f4f7b880-3de2-11ea-866f-c29ff13c6d10.png)

**Comment by ijjk** [comment]
Hi, I updated the example to support nested lists and added an example of it being used on [this post](https://notion-blog.now.sh/blog/my-second-post)

**Comment by dnnaji** [comment]
Hi JJ,
Would you be able to share you high level thinking on how you tackled this development?

I was looking into this but didn't know where to start. Your solution make sense surely, but what would you recommend for someone else to be able to solve similar challenges? Learn more React? JS? OR just programming in general?

**Comment by ijjk** [comment]
Hi @dnnaji, to solve it I initially looked out what the requirements of the data structure would be to decide which structure would be ideal to use. Since it's not a normal list `[1, 2, 3]` and items can be nested it made sense to use a [linked list](https://en.wikipedia.org/wiki/Linked_list). 

Normally a linked list is implemented by pointing to the reference of the next item but to speed up the look up by preventing having to iterate through the list to find the parent reference I implemented a flat linked list where all nodes are available at the top of the map by their id so I could look up any parent's at the top level of the map by their id and add the nested item onto it's nested list. 

Then to render the structure I used my knowledge of how React [creates elements](https://reactjs.org/docs/react-api.html#createelement) and [recursion](https://en.wikipedia.org/wiki/Recursion_(computer_science)) to get all of the top level list items and then recursively render their nested items.

I think having a solid understanding of programming structures/algorithms can help a lot and then knowing enough JS to implement those structures/algorithms can make solving problems a lot more approachable. Once you understand those, understanding React should be more approachable also.

I hope this helped! Let me know if you have any follow up questions.
